File: elixir/triangle.ex
defmodule Triangle do
  @type kind :: :equilateral | :isosceles | :scalene

  @doc """
  Return the kind of triangle of a triangle with 'a', 'b' and 'c' as lengths.
  """
  @spec kind(number, number, number) :: {:ok, kind} | {:error, String.t()}
  def kind(a, b, c) do
    [a, b, c] = Enum.sort([a, b, c])

    cond do
      a <= 0 || b <= 0 || c <= 0 -> {:error, "all side lengths must be positive"}
      a + b <= c -> {:error, "side lengths violate triangle inequality"}
      a == b and b == c -> {:ok, :equilateral}
      a == b || b == c -> {:ok, :isosceles}
      true -> {:ok, :scalene}
    end
  end
end

File: prolog/triangle.pl
is_triangle(Side1, Side2, Side3) :-
    Side1 > 0,
    Side2 > 0,
    Side3 > 0,
    Side1 + Side2 >= Side3,
    Side2 + Side3 >= Side1,
    Side1 + Side3 >= Side2.

triangle(Side1, Side2, Side3, "equilateral") :-
    is_triangle(Side1, Side2, Side3),
    Side1 = Side2,
    Side1 = Side3.

triangle(Side1, Side2, Side3, "isosceles") :-
    is_triangle(Side1, Side2, Side3),
    (Side1 = Side2 ; Side1 = Side3 ; Side2 = Side3).

triangle(Side1, Side2, Side3, "scalene") :-
    is_triangle(Side1, Side2, Side3),
    \+ triangle(Side1, Side2, Side3, "isosceles").
